### ISSUE TITLE
pasting code loses formatting

### ISSUE THREAD

**Issue by cdepillabout** [idris-hackers/idris-mode#457]
I'm having a problem with copying and pasting code.

Let's say I have the following function:

```idris
getEntry : (pos : Integer) -> (store : DataStore) ->
           Maybe (String, DataStore)
getEntry pos dataStore@(MkData schema' size' store') =
  case integerToFin pos size' of
    Nothing => Just ("Out of range\n", dataStore)
    Just id' => Just (?display (index id' store'), dataStore)
```

If I highlight this and then go to another part of the file and try to paste it, it comes out looking like this:

```idris
getEntry : (pos : Integer) -> (store : DataStore) ->
Maybe (String, DataStore)
getEntry pos dataStore@(MkData schema' size' store') =
case integerToFin pos size' of
Nothing => Just ("Out of range\n", dataStore)
Just id' => Just (?display (index id' store'), dataStore)
```

All formatting has been lost.

Is this a problem with idris-mode?

I'm relatively new to emacs, and I'm using [spacemacs](https://github.com/syl20bnr/spacemacs) with evil-mode.  It is entirely possible that this is a problem with spacemacs or evil-mode (or the [idris](https://github.com/syl20bnr/spacemacs/tree/master/layers/%2Blang/idris) layer in spacemacs), and *not* idris-mode.

However, I don't  have this problem in other major modes (including when editing Haskell files using the [haskell](https://github.com/syl20bnr/spacemacs/tree/master/layers/%2Blang/haskell) layer), so I'm thinking that this might be a problem with idris-mode.

On the other hand, no one else has posted about this, so it may just be a problem with my setup.  I'm not sure if anyone who would see this issue is using spacemacs, but just in case here is my [.spacemacs](https://github.com/cdepillabout/docs/blob/master/dot_files/dot_spacemacs).

**Comment by cdepillabout** [comment]
My temporary workaround is to switch to fundamental-mode (with `SPC SPC fundamental-mode`) and then switch back to idris-mode (with `SPC SPC idris-mode`).

**Comment by cdepillabout** [comment]
It seems like this might only be occurring if the buffer has already been loaded into the repl with `idris-load-file`.

If the buffer has *not yet* been loaded into the repl, then pasting seems to work normally. 

I wonder if this has something to do with the semantic highlighting?

**Comment by david-christiansen** [comment]
I can't replicate this on my computer and Emacs setup.

I wasn't able to test with your particular file, as you didn't include enough of it to get it to type check and highlight, so I used the following:
```
module Test

import Data.Vect


foo : (n : Nat) -> Vect n ()
foo n =
    case n of
      Z   => []
      S k => () :: foo k
```

Using a fairly ordinary Emacs setup, I first used the mouse to highlight the whole definition of `foo`, then did `M-w` (`kill-ring-save`) to put it on the kill ring, and then pressed `RET`  a few times to insert some empty lines, and then did `C-y` (`yank`) to paste it in. The result was as expected - a non-highlighted copy of the definition, indented correctly.

Which underlying Emacs commands are your copy-paste keybindings running?

**Comment by cdepillabout** [comment]
Thanks for your response @david-christiansen.  It helped me pinpoint where the error is actually occuring.

I tried directly using the commands you suggested (`kill-ring-save` and `yank`), but the formatting was still lost.

This led me to believe that the problem was actually in spacemacs and **not** in idris-mode.

Through a little trial-and-error, I was able to figure out a solution (work-around?).  I posted my solution as an issue to the spacemacs repo:

https://github.com/syl20bnr/spacemacs/issues/9433

Thanks for your help.  Since this doesn't seem to be a problem with idris-mode, I'll go ahead and close this issue.
